docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,157 @@
+
+The cluster node adding the file will read in a stream of bytes that are then passed to an importer to create an ipfs files dag chunked (size-splitter or rabin-splitter) and layed out (trickle or balanced) per user instruction.  Before importing blocks of each shard the cluster peer calls the `allocate()` function to find the pinning peer for the upcoming shard.  Concurrent with ipfs files dag creation and transmission of the dag's blocks to the allocated cluster peer, the importing ipfs-cluster node will oversee construction of a cluster-dag shard node using the cids of the blocks of the ipfs files dag.  Once a shard is filled with cid references it will also be transmitted and pinned to the allocated ipfs-cluster node.  This process continues until both dags are fully constructed and pinned as expected.  The importing node transmits blocks by calling `ipfs block put` and `ipfs pin` on the pinning node's ipfs daemon via an RPC to the pinning cluster peer.   The pinning node should check the state to ensure it was actually assigned the pin before running the rpc to completion.  Note that this will be one of the more involved parts of implementing basic sharding and will involve extracting, refactoring and adding features to the dag building functionality of go-ipfs.
+
+There will likely be several endpoints entering into the functionality for adding sharded data.  For example, there will be a go endpoint `Add(r io.Reader, opts...)` which just adds arbitrary data. There will be an HTTP endpoint to which users can post multipart data just like it is now posted to ipfs and which creates the unixfs merkle dag.  If dags can be streamed node by node through ipfs safely without possibly overloading the process and its repo, cluster will likely support an AddDag(chan <-*dagNode) endpoint which allocates already-imported-dags.
+
+We will likely abstract the importing functionality into an importer component of cluster.  Following current conventions other cluster modules would communicate with this component over RPC.  However, the RPC module we use does not handle transmitting streams of data so this component and the components containing the endpoints described above will need to expose a stream friendly abstraction, potentially an HTTP endpoint or socket.
```

Comment:
From here til the end everything is in cursive... intended?

Reply:
No, but if you view as markdown the effects go away.  I'll see if I can fix the raw text so github doesn't do this.

Reply:
it comes from chan <-*dagNode. I guess just quoting that will help.